generic/MaplGenericComponent.F90

```diff
@@ -167,4 +219,254 @@ subroutine set_logger(this, lgr)
+     integer :: num_children, i, status
+      
+     this%threading_active = .TRUE.
+     num_children = this%get_num_children()
+
+     if (.NOT. allocated(this%subcomponents)) then
+        call this%create_subobjects(num_threads, __RC__)
+     end if
+
+     do i = 1, num_children
+        child => this%get_child(i) 
+        SELECT TYPE (child)
+        CLASS IS (MaplGenericComponent)
+           call child%activate_threading(num_threads)
+        CLASS DEFAULT
+           _FAIL('illegal type for child')
+        END SELECT
+     end do
+     _RETURN(0)
+   end subroutine activate_threading
+
+   subroutine create_subobjects(this, num_threads, unusable, rc)
+     class(MaplGenericComponent), intent(inout) :: this
+     integer, intent(in) :: num_threads
+     class(KeywordEnforcer), optional, intent(in) :: unusable
+     integer, optional, intent(out) :: rc
+     integer :: i, status
+     type(ESMF_Grid), allocatable :: subgrids(:)
+
+     allocate(this%subcomponents(num_threads))
+     
+     this%subcomponents(:)%import_state = make_substates(this%import_state, num_threads, __RC__)
+     this%subcomponents(:)%export_state = make_substates(this%export_state, num_threads, __RC__)
+     this%subcomponents(:)%internal_state = make_substates(this%internal_state, num_threads, __RC__)
+
+     subgrids = make_subgrids(this%grid%ESMFGrid, num_threads, __RC__) ! make_subgrids requires grid of type ESMF_Grid
+     do i = 1, size(subgrids)
+        call this%subcomponents(i)%grid%set(subgrids(i), __RC__)
+     end do
+     deallocate(subgrids)
+
+     this%subcomponents(:)%gridcomp = this%make_subgridcomps(num_threads, __RC__)
+
+     _RETURN(0)
+   end subroutine create_subobjects
+
+   subroutine deactivate_threading(this, unusable, rc)
+     class(MaplGenericComponent), intent(inout) :: this
+     class(KeywordEnforcer), optional :: unusable
+     integer, optional, intent(out) :: rc
+     class(AbstractFrameworkComponent), pointer :: child
+     integer :: num_children, i
+     
+     num_children = this%get_num_children()
+     do i = 1, num_children
+        child => this%get_child(i) 
+        SELECT TYPE (child)
+        CLASS IS (MaplGenericComponent)
+           call child%deactivate_threading()
+        CLASS DEFAULT
+           _FAIL('illegal type for child')
+        END SELECT
+     end do
+     
+     this%threading_active = .FALSE.
+     _RETURN(0)
+   end subroutine deactivate_threading
+
+
+
+   function get_gridcomp(this) result(gridcomp)
+      class(MaplGenericComponent), target, intent(in) :: this
+      type(ESMF_GridComp), pointer :: gridcomp
+
+      integer :: thread 
+
+      if (.NOT. this%is_threading_active()) then
+         gridcomp => this%gridcomp
+     else
+        thread = 0
+        !$ thread = omp_get_thread_num()
+        gridcomp => this%subcomponents(thread+1)%gridcomp
+     end if
+
+  end function get_gridcomp
+
+  function make_subgridcomps(this, num_grids, unusable, rc) result(subgridcomps)
```

Comment:
I'm sure there is a reason that this procedure cannot go into `OpenMP_Support.F90` with the other "sub" creator procedures.  But please double check just in case it can.